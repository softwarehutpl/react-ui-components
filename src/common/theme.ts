const theme: { [key: string]: { [key: string]: { [key: string]: string } } } = {
  colors: {
    primary: {
      base: '#000000',
      light: '#ffffff',
      dark: '#a8b3bd',
    },
    secondary: {
      base: '#3690e3',
      light: '#75b0e6',
      dark: '#034d91',
    },
    error: {
      base: '#f44336',
      light: '#f56c62',
      dark: '#a1140a',
    },
    warning: {
      base: '#faa72d',
      light: '#ffc36b',
      dark: '#995b00',
    },
    info: {
      base: '#7979d4',
      light: '#9797d1',
      dark: '#4c4cc7',
    },
    success: {
      base: '#4caf50',
      light: '#88c28a',
      dark: '#124f14',
    },
    disabled: {
      base: '#d9d9d9',
      light: '#f0eded',
      dark: '#8c8989',
    },
  },
  // TODO: change it in dedicated task for that (change ease-out for example and add more options)
  transitions: {
    all: {
      fast: 'all 0.5s ease-out',
      mid: 'all 2s ease-out',
      slow: 'all 4s ease-out',
    },
    background: {
      fast: 'background-color 0.5s ease-out',
      mid: 'background-color 2s ease-out',
      slow: 'background-color 4s ease-out',
    },
    fontColor: {
      fast: 'color 0.5s ease-out',
      mid: 'color 2s ease-out',
      slow: 'color 4s ease-out',
    }
  }
};

export default theme;
