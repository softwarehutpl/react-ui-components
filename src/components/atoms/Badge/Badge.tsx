import React from 'react';
import styled, { css } from 'styled-components';

export type Position = {
  horizontal?: any;
  vertical?: any;
};

export interface BadgeWrapperProps {
  margin?: number;
  padding?: number
};

export interface BadgeCircleProps {
  color?: string;
  backgroundColor?: string;
  fontColor?: string;
  width?: number;
  height?: number;
  position?: Position;
};

export interface BadgeContentProps {
  max?: number;
};

export interface BadgeProps extends BadgeWrapperProps, BadgeCircleProps, BadgeContentProps {
  className?: string;
  badgeContent?: any;
  children?: any;
};

export const BadgeWrapper = styled.div<BadgeWrapperProps>`
  position: relative;
  display: inline-flex;
  flex-shrink: 0;
  vertical-align: middle;
  margin: ${({ margin }) => margin + 'px'};
  padding: ${({ padding }) => padding + 'px'};
`;

export const BadgeCircle = styled.span<BadgeCircleProps>`
  background-color: ${({ backgroundColor, color, theme}) => (backgroundColor || (color && theme.colors[color].base))};
  color: ${({ fontColor, color, theme }) =>(fontColor || (color && theme.colors[color].light))};
  width: ${({ width }) => (width  &&`${width}px`)};
  height: ${({ height }) => (height && `${height}px`)};
  border-radius: 10px;
  padding: 0 7px;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
  font-size: 0.75rem;
  position: absolute;
  line-height: 1;
  transform: scale(1) translate(50%, -50%);
  transform-origin: 100% 0%;

  ${({ position }) => 
    position  && css`
     top: ${() => (position.vertical === 'top' && '0')};
     bottom: ${() => (position.vertical === 'bottom' && '-15px')};
     right: ${() => (position.horizontal === 'right' && '0')};
     left: ${() => (position.horizontal === 'left' && '-20px')};
    `
  }
`;

export const BadgeText = styled.div``
export const BadgeContent = styled.div`
  width: 100%;
  height: 100%;
`;

 const Badge: React.StatelessComponent<BadgeProps> = ({ 
   className, 
   margin,
   padding,
   badgeContent,
   children,
   color,
   backgroundColor,
   fontColor,
   width,
   height,
   position,
   max
  }) => (
    <BadgeWrapper
      className={className}
      margin={margin}
      padding={padding}
    >
      <BadgeCircle 
        color={color}
        backgroundColor={backgroundColor}
        fontColor={fontColor}
        width={width}
        height={height}
        position={position}
      >
        {typeof badgeContent === 'number' && max ? `${(badgeContent - 1).toString()} +` : badgeContent}
      </BadgeCircle>
      <BadgeContent>{children}</BadgeContent>
    </BadgeWrapper>
  );

const defaultProps = {
  className: '',
  color: 'primary',
  height: 20,
  position: {
    vertical: 'top',
    horizontal: 'right',
  },
  margin: 25,
  padding: 10,
  max: 0,
};

Badge.defaultProps = defaultProps;

export default Badge;
