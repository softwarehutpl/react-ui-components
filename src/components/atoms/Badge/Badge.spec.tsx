import React from 'react';
import { create, act } from 'react-test-renderer';
import 'jest-styled-components';
import Badge, { BadgeWrapper, BadgeCircle } from './Badge';
import theme from '../../../common/theme'


describe('<Badge', () => {
  it('should render corectly', () => {
    const component = create(<BadgeWrapper theme={theme} />);
    expect(component).toMatchSnapshot();
  });

  it('should render with custom class', () => {
    const component = create(<BadgeWrapper theme={theme} className="test-class" />).toJSON();
    expect(component.props.className).toEqual(expect.stringContaining('test-class'));
  });

  it('should render with proper margin and padding style if provided', () => {
    const component = create(<BadgeWrapper theme={theme} margin={20} padding={10} />).toJSON();
    expect(component).toHaveStyleRule('margin', '20px');
    expect(component).toHaveStyleRule('padding', '10px');
  });

  it('should render with default background color and font color', () => {
    const component = create(<BadgeCircle theme={theme} color="primary" />).toJSON();
    expect(component).toHaveStyleRule('background-color', '#000000');
    expect(component).toHaveStyleRule('color', '#ffffff');
  });

  it('should render with proper backgroundColor and fontColor if provied', () => {
    const component = create(<BadgeCircle theme={theme} backgroundColor="hotpink" fontColor="#fff" />).toJSON();
    expect(component).toHaveStyleRule('background-color', 'hotpink');
    expect(component).toHaveStyleRule('color', '#fff');
  });

  it('should render with proper horizontal and vertical position', () => {
    const component = create(<BadgeCircle theme={theme} position={{horizontal: 'left', vertical: 'bottom'}} />).toJSON();
    expect(component).toHaveStyleRule('left', '-20px');
    expect(component).toHaveStyleRule('bottom', '-15px');
  });
});
