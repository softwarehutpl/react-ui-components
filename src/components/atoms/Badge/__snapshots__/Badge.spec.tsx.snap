// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Badge should render corectly 1`] = `
.c0 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  vertical-align: middle;
  margin: undefinedpx;
  padding: undefinedpx;
}

<div
  className="c0"
/>
`;
